### PR TITLE
Persist PNC HF next visit date

### PR DESCRIPTION
**Pull request by allan-allay**
- Added JSON form field for PNC HF next visit date
- Added DB migrations to process home visits and get next visit dates

### CODE REVIEW

File: opensrp-chw/src/ba/assets/json.form/pnc_health_facility_visit.json [hunk 1]
```diff
@@ -155,6 +155,13 @@
         "read_only": "true",
         "hidden": "true",
         "value": "0"
+      },
+      {
+        "key": "pnc_hf_next_visit_date",
+        "openmrs_entity_parent": "",
+        "openmrs_entity": "concept",
+        "openmrs_entity_id": "anc_hf_next_visit_date",
```
Comment (rkodev) [inline opensrp-chw/src/ba/assets/json.form/pnc_health_facility_visit.json]: Should you be using **anc_hf_next_visit_date** as the entity ID? This should be the **pnc_hf_next_visit_date**

File: opensrp-chw/src/ba/assets/json.form/pnc_health_facility_visit_sw.json [hunk 2]
```diff
@@ -167,6 +167,13 @@
         "read_only": "true",
         "hidden": "true",
         "value": "0"
+      },
+      {
+        "key": "pnc_hf_next_visit_date",
```
Comment (rkodev) [inline opensrp-chw/src/ba/assets/json.form/pnc_health_facility_visit_sw.json]: Should you be using **anc_hf_next_visit_date** as the entity ID? This should be the **pnc_hf_next_visit_date**

File: opensrp-chw/src/chad/assets/json.form/pnc_health_facility_visit.json [hunk 3]
```diff
@@ -167,6 +167,13 @@
         "read_only": "true",
         "hidden": "true",
         "value": "0"
+      },
+      {
+        "key": "pnc_hf_next_visit_date",
```
Comment (rkodev) [inline opensrp-chw/src/chad/assets/json.form/pnc_health_facility_visit.json]: The same issue looks like a copy-paste on all the other fields

File: opensrp-chw/src/main/java/org/smartregister/chw/interactor/DefaultPncHomeVisitInteractorFlv.java [hunk 4]
```diff
@@ -677,6 +677,10 @@ public String getPreProcessed() {
                 pnc_visit_date.put("hint",
                         MessageFormat.format(pnc_visit_date.getString(JsonFormConstants.HINT), visitRule.getVisitName())
                 );
+
+                JSONObject pnc_hf_next_visit_date = JsonFormUtils.getFieldJSONObject(fields, "pnc_hf_next_visit_date");
+                pnc_hf_next_visit_date.put(JsonFormConstants.VALUE, DateTimeFormat.forPattern("dd-MM-yyyy").print(visitRule.getDueDate()));
```
Comment (rkodev) [inline opensrp-chw/src/main/java/org/smartregister/chw/interactor/DefaultPncHomeVisitInteractorFlv.java]: This is a bug, the visit due date you are using is for the current visit, not the next visit

File: opensrp-chw/src/wcaro/assets/json.form/pnc_health_facility_visit_fr.json [hunk 5]
```diff
@@ -171,6 +171,13 @@
         "read_only": "true",
         "hidden": "true",
         "value": "0"
+      },
```
Comment (rkodev) [inline opensrp-chw/src/wcaro/assets/json.form/pnc_health_facility_visit_fr.json]: same issue on naming

File: opensrp-chw/src/wcaro/assets/json.form/pnc_health_facility_visit.json [hunk 6]
```diff
@@ -167,6 +167,13 @@
         "read_only": "true",
         "hidden": "true",
         "value": "0"
+      },
```
Comment (rkodev) [inline opensrp-chw/src/wcaro/assets/json.form/pnc_health_facility_visit.json]: same issue on naming

File: opensrp-chw/src/togo/assets/json.form/pnc_health_facility_visit_fr.json [hunk 7]
```diff
@@ -171,6 +171,13 @@
         "read_only": "true",
         "hidden": "true",
         "value": "0"
+      },
```
Comment (rkodev) [inline opensrp-chw/src/togo/assets/json.form/pnc_health_facility_visit_fr.json]: same issue on naming

File: opensrp-chw/src/togo/assets/json.form/pnc_health_facility_visit.json [hunk 8]
```diff
@@ -167,6 +167,13 @@
         "read_only": "true",
         "hidden": "true",
         "value": "0"
+      },
```
Comment (rkodev) [inline opensrp-chw/src/togo/assets/json.form/pnc_health_facility_visit.json]: same issue on naming

File: opensrp-chw/src/guinea/assets/json.form/pnc_health_facility_visit_fr.json [hunk 9]
```diff
@@ -171,6 +171,13 @@
         "read_only": "true",
         "hidden": "true",
         "value": "0"
+      },
```
Comment (rkodev) [inline opensrp-chw/src/guinea/assets/json.form/pnc_health_facility_visit_fr.json]: same issue on naming

File: opensrp-chw/src/drc/assets/json.form/pnc_health_facility_visit_fr.json [hunk 10]
```diff
@@ -171,6 +171,13 @@
         "read_only": "true",
         "hidden": "true",
         "value": "0"
+      },
+      {
+        "key": "pnc_hf_next_visit_date",
```
Comment (rkodev) [inline opensrp-chw/src/drc/assets/json.form/pnc_health_facility_visit_fr.json]: same issue on naming

File: opensrp-chw/src/drc/assets/json.form/pnc_health_facility_visit.json [hunk 11]
```diff
@@ -167,6 +167,13 @@
         "read_only": "true",
         "hidden": "true",
         "value": "0"
+      },
+      {
+        "key": "pnc_hf_next_visit_date",
```
Comment (rkodev) [inline opensrp-chw/src/drc/assets/json.form/pnc_health_facility_visit.json]: same issue on naming

File: opensrp-chw/src/main/java/org/smartregister/chw/interactor/DefaultPncHomeVisitInteractorFlv.java (hunk 4 above)
Comment (allan-allay) [reply opensrp-chw/src/main/java/org/smartregister/chw/interactor/DefaultPncHomeVisitInteractorFlv.java]: The `visitRule` from which the due date is retrieved is for the next visit and not the current one. 
In line 148 `PNCHealthFacilityVisitRule visitRule = PNCVisitUtil.getNextPNCHealthFacilityVisit(summary.getDeliveryDate(), summary.getLastVisitDate())`

File: opensrp-chw/src/ba/assets/json.form/pnc_health_facility_visit.json (hunk 1 above)
Comment (allan-allay) [reply opensrp-chw/src/ba/assets/json.form/pnc_health_facility_visit.json]: I'll remove the anc_hf_next_visit_date. Wrongly copied this over from the anc_hf_next_visit_date

File: opensrp-chw/src/drc/assets/json.form/pnc_health_facility_visit_fr.json (hunk 10 above)
Comment (allan-allay) [reply opensrp-chw/src/drc/assets/json.form/pnc_health_facility_visit_fr.json]: Fixed as mentioned above

File: opensrp-chw/src/guinea/assets/json.form/pnc_health_facility_visit_fr.json (hunk 9 above)
Comment (allan-allay) [reply opensrp-chw/src/guinea/assets/json.form/pnc_health_facility_visit_fr.json]: Fixed as mentioned above

File: opensrp-chw/src/togo/assets/json.form/pnc_health_facility_visit.json (hunk 8 above)
Comment (allan-allay) [reply opensrp-chw/src/togo/assets/json.form/pnc_health_facility_visit.json]: Fixed as mentioned above

File: opensrp-chw/src/togo/assets/json.form/pnc_health_facility_visit_fr.json (hunk 7 above)
Comment (allan-allay) [reply opensrp-chw/src/togo/assets/json.form/pnc_health_facility_visit_fr.json]: Fixed as mentioned above

File: opensrp-chw/src/wcaro/assets/json.form/pnc_health_facility_visit_fr.json (hunk 5 above)
Comment (allan-allay) [reply opensrp-chw/src/wcaro/assets/json.form/pnc_health_facility_visit_fr.json]: Fixed as mentioned above

File: opensrp-chw/src/ba/assets/json.form/pnc_health_facility_visit_sw.json (hunk 2 above)
Comment (allan-allay) [reply opensrp-chw/src/ba/assets/json.form/pnc_health_facility_visit_sw.json]: Fixed as mentioned above

File: opensrp-chw/src/drc/assets/json.form/pnc_health_facility_visit.json (hunk 11 above)
Comment (allan-allay) [reply opensrp-chw/src/drc/assets/json.form/pnc_health_facility_visit.json]: Fixed as mentioned above

File: opensrp-chw/src/chad/assets/json.form/pnc_health_facility_visit.json (hunk 3 above)
Comment (allan-allay) [reply opensrp-chw/src/chad/assets/json.form/pnc_health_facility_visit.json]: Fixed as mentioned above

**Pull request merged**